### PR TITLE
make send snapshots async

### PR DESCRIPTION
**Pull request by hicqu**
Previously snapshots are sent in a `ThreadPool`, with such defect: if the receiver of one snapshot suspend, the sender will keep blocking in the pool.
This PR make sending snapshots async in a `CpuPool`. And we can make *recveive snapshots* reuse the pool later.

### CODE REVIEW

**Comment by hicqu** [comment]
@BusyJay @overvenus @zhangjinpeng1987 PTAL thanks.

**Comment by hicqu** [comment]
friendly ping @overvenus @BusyJay .

File: src/server/snap.rs [hunk 1]
```diff
@@ -39,24 +38,29 @@ use super::transport::RaftStoreRouter;
 pub type Callback = Box<FnBox(Result<()>) + Send>;
 
 const DEFAULT_SENDER_POOL_SIZE: usize = 3;
+// How many snapshots can be sent concurrent.
```
Comment (zhangjinpeng1987) [inline src/server/snap.rs]: concurrently.

File: src/server/snap.rs [hunk 2]
```diff
@@ -110,96 +137,164 @@ impl Stream for SnapChunk {
...
-    env: Arc<Environment>,
-    mgr: SnapManager,
-    security_mgr: Arc<SecurityManager>,
     addr: &str,
     msg: RaftMessage,
-) -> Result<()> {
+    env: Arc<Environment>,
+    security_mgr: Arc<SecurityManager>,
+    mgr: SnapManager,
+) -> Box<Future<Item = (), Error = Error> + Send> {
```
Comment (siddontang) [inline src/server/snap.rs]: can we use impl Traits to avoid Box here?

Comment (hicqu) [reply src/server/snap.rs]: This function won't be called frequently. And if we write it as `and_then` chain style, we need to clone many `Arc` or similar things. So I think readability is more important here.

Comment (BusyJay) [reply src/server/snap.rs]: Why not use `Result<impl Future>`? So that you can use both `?` and `and_then`.

File: src/server/snap.rs [hunk 3]
```diff
@@ -110,96 +137,164 @@ impl Stream for SnapChunk {
...
-                "[region {}] sent snapshot {} [size: {}, dur: {:?}]",
-                key.region_id,
-                key,
-                total_size,
-                timer.elapsed()
-            );
-            s.wl().delete();
-            Ok(())
-        })
-        .wait()
```
Comment (BusyJay) [inline src/server/snap.rs]: Looks like changing `wait` to `then` can achieve the goal with less changes.

File: src/server/snap.rs [hunk 4]
```diff
@@ -299,17 +316,30 @@ impl<R: RaftStoreRouter + 'static> Runnable<Task> for Runner<R> {
                 }
             }
             Task::SendTo { addr, msg, cb } => {
+                if self.concurrent_send_count.load(Ordering::Acquire) >= MAX_SENDER_CONCURRENT {
+                    info!("drop SendTo Snap[to: {}, snap: {:?}] silently", addr, msg);
+                    return;
```
Comment (BusyJay) [inline src/server/snap.rs]: Callback should be called before returned.

File: src/server/snap.rs [hunk 5]
```diff
@@ -299,17 +316,30 @@ impl<R: RaftStoreRouter + 'static> Runnable<Task> for Runner<R> {
...
                 let security_mgr = Arc::clone(&self.security_mgr);
-                self.pool.execute(move |_| {
-                    let res = send_snap(env, mgr, security_mgr, &addr, msg);
-                    if res.is_err() {
-                        error!("failed to send snap to {}: {:?}", addr, res);
-                    }
-                    cb(res)
-                });
+                let concurrent_send_count = Arc::clone(&self.concurrent_send_count);
+                concurrent_send_count.fetch_add(1, Ordering::Release);
```
Comment (BusyJay) [inline src/server/snap.rs]: Use `SeqCst` instead.

File: src/server/snap.rs [hunk 6]
```diff
@@ -154,34 +160,43 @@ fn send_snap(
...
 
 pub struct Runner<R: RaftStoreRouter + 'static> {
     env: Arc<Environment>,
     snap_mgr: SnapManager,
     files: HashMap<Token, (Box<Snapshot>, RaftMessage)>,
-    pool: ThreadPool<DefaultContext>,
+    pool: CpuPool,
     raft_router: R,
     security_mgr: Arc<SecurityManager>,
+    concurrent_send_count: Arc<AtomicUsize>,
```
Comment (BusyJay) [inline src/server/snap.rs]: s/concurrent_send_count/sending_count/

File: src/server/snap.rs [hunk 7]
```diff
@@ -125,63 +138,64 @@ fn send_snap(
...
+            drop(deregister);
+            drop(client);
+            result.map(|s| {
+                s.snap.delete();
+                // TODO: improve it after rustc resolves the bug.
+                // Call `info` in the closure directly will cause rustc
+                // panic with `Cannot create local mono-item for DefId`.
+                SendStat {
+                    key: key,
+                    total_size,
```
Comment (BusyJay) [inline src/server/snap.rs]: Keep style consistency with other fields.

File: src/server/snap.rs [hunk 8]
```diff
@@ -125,63 +138,64 @@ fn send_snap(
...
-    let s = Arc::new(RwLock::new(s));
-
     let chunks = {
-        let snap_chunk = SnapChunk {
-            snap: Arc::clone(&s),
+        let mut first_chunk = SnapshotChunk::new();
+        first_chunk.set_message(msg);
+
+        SnapChunk {
+            first: Some((first_chunk, WriteFlags::default().buffer_hint(true))),
```
Comment (BusyJay) [inline src/server/snap.rs]: Write flag can be set at L92.

File: src/server/snap.rs [hunk 9]
```diff
@@ -299,17 +315,40 @@ impl<R: RaftStoreRouter + 'static> Runnable<Task> for Runner<R> {
                 }
             }
             Task::SendTo { addr, msg, cb } => {
+                if self.sending_count.load(Ordering::Acquire) >= MAX_SENDER_CONCURRENT {
```
Comment (BusyJay) [inline src/server/snap.rs]: Use `SeqCst` instead.

Comment (hicqu) [reply src/server/snap.rs]: I think `Acquire` and `Release` pair is more relaxed than `SeqCst`?

File: src/server/snap.rs [hunk 10]
```diff
@@ -299,17 +316,40 @@ impl<R: RaftStoreRouter + 'static> Runnable<Task> for Runner<R> {
                 }
             }
             Task::SendTo { addr, msg, cb } => {
+                if self.sending_count.load(Ordering::SeqCst) >= MAX_SENDER_CONCURRENT {
+                    info!("drop SendTo Snap[to: {}, snap: {:?}] silently", addr, msg);
```
Comment (BusyJay) [inline src/server/snap.rs]: Confusing log.

File: src/server/snap.rs [hunk 11]
```diff
@@ -299,17 +316,43 @@ impl<R: RaftStoreRouter + 'static> Runnable<Task> for Runner<R> {
                 }
             }
             Task::SendTo { addr, msg, cb } => {
+                if self.sending_count.load(Ordering::SeqCst) >= MAX_SENDER_CONCURRENT {
+                    info!(
```
Comment (overvenus) [inline src/server/snap.rs]: Please warn.

**Pull request merged**